src/amplitude-client.js

```diff
@@ -126,6 +126,7 @@ AmplitudeClient.prototype.init = function init(apiKey, opt_userId, opt_config, o
       secure: this.options.secureCookie,
       sameSite: this.options.sameSiteCookie,
     });
+    this.options.domain = this.cookieStorage.options().domain;
```

Comment:
im curious how this impacts sdk usages that did not use defer init. would they have existing cookie without the domain suffix? if we move it up to this line, then would it look for cookie with the domain prefix, which it may not find?

Reply:
It is really interesting that we actually have two cookie names: with domain and without domain. 
![image](https://user-images.githubusercontent.com/5042222/167509668-bf224580-aa6a-4001-b1b9-dec6fec63ecc.png)
![image](https://user-images.githubusercontent.com/5042222/167509703-4628febe-ea44-4a09-acef-0a7a02efc430.png)

And it does not defer init
![image](https://user-images.githubusercontent.com/5042222/167510015-b867d72e-6ef0-4e06-8780-c061f13c3920.png)

Reply:
When initializing the metadata storage without domain, all cookies saved by the metadata storage does not include the domain suffix. After domain is assigned to `this.options.domain`, SDK looks for cookie but cannot find it, and then a new cookies with domain suffix is created. I think this is the flow.

Reply:
Hmm not sure about that, but isn't metadata storage only instantiated once for the most part? then does this mean even if `this.options.domain` changes, it would still use the initial storage key?

Reply:
> isn't metadata storage only instantiated once for the most part?

If deferInitialization is not set, then it is only one metadata storage instance. But for defer init, a new metadata storage instance is initialized with the new domain value. 


> then does this mean even if `this.options.domain` changes, it would still use the initial storage key?

If `this.options.domain` changes, the storage key is still the initial one. 

Here we have two options:
1) Move the `this.options.domain = this.cookieStorage.options().domain;` before initialize metadata storage, so the cookies include domain suffix.
2) Move the `this.options.domain = this.cookieStorage.options().domain;` after the defer init check, so the metadata storage is initialized again, the storage key is still the same. 

I pick the first one option, but I am thinking if you have any other concern.

Reply:
From `setDomain` method, the `this.options.domain` is updated immediately. 

https://github.com/amplitude/Amplitude-JavaScript/blob/52abaf062492217ddf8d556b85509ce00e006f25/src/amplitude-client.js#L862-L880

Reply:
Yeah my concern is what I mentioned in my first comment. Like if they don't defer init, which is default behavior, and client devices have an existing cookie key with no suffix. If we go with option 1, then it will not find the existing cookies right? If it does not find existing cookies, then it will generate new session ids and device ids and potentially disassociate events that would have been for the same amplitude ID, I'm thinking.

Reply:
Right. I think it is a valid concern. I will move to option 2.